druid/src/widget/stepper.rs

```diff
@@ -0,0 +1,208 @@
+        if self.increase_active {
+            paint_ctx.fill(increase_button_rect, &active_gradient);
+        } else {
+            paint_ctx.fill(increase_button_rect, &inactive_gradient);
+        };
+
+        if self.decrease_active {
+            paint_ctx.fill(decrease_button_rect, &active_gradient);
+        } else {
+            paint_ctx.fill(decrease_button_rect, &inactive_gradient);
+        };
+
+        // draw up and down triangles
+        let mut increase_button_arrow = BezPath::new();
+        increase_button_arrow.move_to(Point::new(4., height / 2. - 4.));
+        increase_button_arrow.line_to(Point::new(width - 4., height / 2. - 4.));
+        increase_button_arrow.line_to(Point::new(width / 2., 4.));
+        increase_button_arrow.close_path();
+        paint_ctx.fill(increase_button_arrow, &env.get(theme::LABEL_COLOR));
+
+        let mut decrease_button_arrow = BezPath::new();
+        decrease_button_arrow.move_to(Point::new(4., height / 2. + 4.));
+        decrease_button_arrow.line_to(Point::new(width - 4., height / 2. + 4.));
+        decrease_button_arrow.line_to(Point::new(width / 2., height - 4.));
+        decrease_button_arrow.close_path();
+        paint_ctx.fill(decrease_button_arrow, &env.get(theme::LABEL_COLOR));
+    }
+
+    fn layout(
+        &mut self,
+        _layout_ctx: &mut LayoutCtx,
+        bc: &BoxConstraints,
+        _data: &f64,
+        env: &Env,
+    ) -> Size {
+        bc.constrain(Size::new(
+            env.get(theme::BASIC_WIDGET_HEIGHT),
+            env.get(theme::BORDERED_WIDGET_HEIGHT),
+        ))
+    }
+
+    fn event(&mut self, ctx: &mut EventCtx, event: &Event, data: &mut f64, env: &Env) {
+        let height = env.get(theme::BORDERED_WIDGET_HEIGHT);
+
+        match event {
+            Event::MouseDown(mouse) => {
+                ctx.set_active(true);
+
+                if mouse.pos.y > height / 2. {
+                    self.decrease_active = true;
+                } else {
+                    self.increase_active = true;
+                }
+
+                self.change_value(ctx, data, env);
+
+                let delay = Instant::now() + Duration::from_millis(500);
+                self.timer_id = ctx.request_timer(delay);
+
+                ctx.invalidate();
+            }
+            Event::MouseUp(_) => {
+                ctx.set_active(false);
+
+                self.decrease_active = false;
+                self.increase_active = false;
+                self.timer_id = TimerToken::INVALID;
+
+                ctx.invalidate();
+            }
+            Event::Timer(id) if *id == self.timer_id => {
+                self.change_value(ctx, data, env);
+                let delay = Instant::now() + Duration::from_millis(200);
+                self.timer_id = ctx.request_timer(delay);
+            }
+            _ => (),
+        }
+    }
+
+    fn update(&mut self, ctx: &mut UpdateCtx, _old_data: Option<&f64>, _data: &f64, _env: &Env) {
+        ctx.invalidate();
```

Comment:
in general in `update` we want to only invalidate if the data has changed. The idiomatic way of writing this would be,

```rust
if !Some(data).same(old_data) {
    ctx.invalidate()
}
```

Reply:
if #454 lands this snippet won't work anymore, my alternative would be:

```rust
if old.map(|old| !old.same(new)).unwrap_or(true) { ... }
```